src/theory/theory.cpp

```diff
@@ -515,16 +515,37 @@ bool Theory::areCareDisequal(TNode x, TNode y)
   Assert(d_equalityEngine != nullptr);
   Assert(d_equalityEngine->hasTerm(x));
   Assert(d_equalityEngine->hasTerm(y));
+  if (x == y)
+  {
+    return false;
+  }
+  if (x.isConst() && y.isConst())
+  {
+    return true;
+  }
   if (!d_equalityEngine->isTriggerTerm(x, d_id)
       || !d_equalityEngine->isTriggerTerm(y, d_id))
   {
+    // just check if they are disequal, which is the used in the case for
+    // non-shared terms.
+    if (d_equalityEngine != nullptr
+        && d_equalityEngine->areDisequal(x, y, false))
+    {
+      return true;
+    }
```

Comment:
You don't need to check d_equalityEngine != nullptr as this is asserted above.

This changes the contract of the function which currently states that it returns true if both terms are shared and disequal.  In this case, one of the terms is not shared.  Are we 100% sure that this won't cause a problem?

Reply:
Addressed.